hydroflow_lang/src/graph/ops/join_fused.rs

```diff
@@ -0,0 +1,362 @@
+            write_iterator_after,
+        })
+    },
+};
+
+pub enum JoinOptions<'a> {
+    FoldFrom(&'a Expr, &'a Expr),
+    Fold(&'a Expr, &'a Expr),
+    Reduce(&'a Expr),
+}
+
+pub fn parse_argument<'a>(
+    diagnostics: &mut Vec<Diagnostic>,
+    arg: &'a Expr,
+) -> Result<JoinOptions<'a>, ()> {
+    let lhs = match arg {
+        Expr::Call(ExprCall {
+            attrs: _,
+            func,
+            paren_token: _,
+            args,
+        }) => {
+            let mut elems = args.iter();
+            let func_name = func.to_token_stream().to_string();
+
+            match func_name.as_str() {
+                "Fold" => match (elems.next(), elems.next()) {
+                    (Some(default), Some(fold)) => JoinOptions::Fold(default, fold),
+                    _ => {
+                        diagnostics.push(Diagnostic::spanned(
+                                args.span(),
+                                Level::Error,
+                                format!("Fold requires two arguments, first is the default function, second is the folding function: {func:?}"),
+                            ));
+                        return Err(());
+                    }
+                },
+                "FoldFrom" => match (elems.next(), elems.next()) {
+                    (Some(from), Some(fold)) => JoinOptions::FoldFrom(from, fold),
+                    _ => {
+                        diagnostics.push(Diagnostic::spanned(
+                                args.span(),
+                                Level::Error,
+                                format!("FoldFrom requires two arguments, first is the From function, second is the folding function: {func:?}"),
+                            ));
+                        return Err(());
+                    }
+                },
+                "Reduce" => match elems.next() {
+                    Some(reduce) => JoinOptions::Reduce(reduce),
+                    _ => {
+                        diagnostics.push(Diagnostic::spanned(
+                            args.span(),
+                            Level::Error,
+                            format!(
+                                "Reduce requires one argument, the reducing function: {func:?}"
+                            ),
+                        ));
+                        return Err(());
+                    }
+                },
+                _ => {
+                    diagnostics.push(Diagnostic::spanned(
+                        func.span(),
+                        Level::Error,
+                        format!("Unknown summarizing function: {func:?}"),
+                    ));
+                    return Err(());
+                }
+            }
+        }
+        _ => {
+            diagnostics.push(Diagnostic::spanned(
+                arg.span(),
+                Level::Error,
+                "argument must be a tuple",
+            ));
+            return Err(());
+        }
+    };
+
+    Ok(lhs)
+}
```

Comment:
Ought to restructure this using let-else, and probably return `Result<JoinOptions<'a>, Diagnostic>` instead of pushing to the diagnostics list directly

Reply:
cool let-else